File: app/src/main/java/com/wezom/kiviremoteserver/service/aspect/AspectLayoutService_NEW.java
//package com.wezom.kiviremoteserver.service.aspect;
//
//
//import android.app.Service;
//import android.content.Context;
//import android.content.Intent;
//import android.graphics.Color;
//import android.graphics.PixelFormat;
//import android.os.Build;
//import android.os.Handler;
//import android.os.IBinder;
//import android.provider.Settings;
//import android.support.annotation.NonNull;
//import android.support.annotation.Nullable;
//import android.support.v7.widget.CardView;
//import android.support.v7.widget.LinearLayoutManager;
//import android.support.v7.widget.RecyclerView;
//import android.util.Log;
//import android.view.KeyEvent;
//import android.view.LayoutInflater;
//import android.view.View;
//import android.view.ViewGroup;
//import android.view.WindowManager;
//import android.view.animation.Animation;
//import android.view.animation.AnimationUtils;
//import android.widget.ImageView;
//import android.widget.RelativeLayout;
//import android.widget.TextView;
//
//import com.wezom.kiviremoteserver.App;
//import com.wezom.kiviremoteserver.R;
//import com.wezom.kiviremoteserver.environment.EnvironmentInputsHelper;
//import com.wezom.kiviremoteserver.environment.EnvironmentPictureSettings;
//import com.wezom.kiviremoteserver.service.aspect.items.IFLMCard;
//import com.wezom.kiviremoteserver.service.aspect.items.InputsCardItem;
//import com.wezom.kiviremoteserver.service.aspect.items.NumericCardItem;
//import com.wezom.kiviremoteserver.service.aspect.items.PictureCardItem;
//import com.wezom.kiviremoteserver.service.aspect.items.RatioCardItem;
//import com.wezom.kiviremoteserver.service.aspect.items.SettingCardItem;
//import com.wezom.kiviremoteserver.service.aspect.items.SoundCardItem;
//import com.wezom.kiviremoteserver.service.aspect.items.TimerCardItem;
//import com.wezom.kiviremoteserver.service.aspect.view.CircleListView;
//import com.wezom.kiviremoteserver.service.aspect.view.IndicatorView;
//
//import java.util.ArrayList;
//import java.util.List;
//
//import wezom.kiviremoteserver.environment.bridge.BridgePicture;
//import wezom.kiviremoteserver.environment.bridge.driver_set.PictureMode;
//import wezom.kiviremoteserver.environment.bridge.driver_set.Ratio;
//import wezom.kiviremoteserver.environment.bridge.driver_set.SoundValues;
//
//
//public class AspectLayoutService_NEW extends Service {
//    public static volatile long lastUpdate;
//    private WindowManager windowManager;
//    private static int mainColor = Color.BLUE;
//    private static final String OSD_TIME = "OSD_TIME";
//
//    private EnvironmentPictureSettings pictureSettings;
//    private EnvironmentInputsHelper inputsHelper;
//
//    private int autoCloseTime = 10;
//    private Handler timer = new Handler();
//    private int generalType = BridgePicture.LAYER_TYPE;//WindowManager.LayoutParams.TYPE_TOAST;
//    private RelativeLayout generalView;
//    RecyclerView mainMenu;
//    CircleListView secondaryMenu;
//    private Runnable updateSleepTime = new Runnable() {
//        @Override
//        public void run() {
////            if (sleepFocused) {
//            // updateSleepText();
////            }
//
//
//            timer.postDelayed(updateSleepTime, 1000);
//            if (autoCloseTime > 0)
//                if (System.currentTimeMillis() - autoCloseTime > lastUpdate) {
//                    stopSelf();
//                }
//        }
//
//
//    };
//
//
//    @Override
//    public void onCreate() {
//        super.onCreate();
//
//        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
//            if (App.checkWizard(this)) {
//                return;
//            }
//        }
//        autoCloseTime = Settings.Global.getInt(getContentResolver(), OSD_TIME, 10);
//
//        if (autoCloseTime < 10 && autoCloseTime > 0) {
//            autoCloseTime = 10;
//        }
//
//        //TODO remove autoCloseTime *= 10;
//        autoCloseTime *= 1000000;
//        autoCloseTime *= 1000;
//        mainColor = getResources().getColor(R.color.colorPrimary);
//
//        lastUpdate = System.currentTimeMillis();
//        pictureSettings = new EnvironmentPictureSettings();
//        inputsHelper = new EnvironmentInputsHelper();
//        createLayout(getBaseContext());
//        timer.postDelayed(updateSleepTime, autoCloseTime);
//
//    }
//
//    @Nullable
//    @Override
//    public IBinder onBind(Intent intent) {
//        return null;
//    }
//
//
//    public void createLayout(Context context) {
//        LayoutInflater layoutInflater = (LayoutInflater)
//                getSystemService(Context.LAYOUT_INFLATER_SERVICE);
//        windowManager = (WindowManager) context.getApplicationContext()
//                .getSystemService(Context.WINDOW_SERVICE);
//        generalView = (RelativeLayout) View.inflate(context, R.layout.layout_aspect_v2, null);
//        final WindowManager.LayoutParams param = new WindowManager.LayoutParams(
//                ViewGroup.LayoutParams.MATCH_PARENT,
//                ViewGroup.LayoutParams.MATCH_PARENT,
//                generalType,//TYPE_SYSTEM_ALERT
//                WindowManager.LayoutParams.FLAG_FULLSCREEN,
//                PixelFormat.TRANSLUCENT);
//        windowManager.addView(generalView, param);
//        generalView.setVisibility(View.VISIBLE);
//        generalView.clearAnimation();
//
//        final Animation anDesk = AnimationUtils.loadAnimation(this, R.anim.outside_bottom);
//        anDesk.setDuration(300);
//        //TODO Animation start
//
//
//        mainMenu = generalView.findViewById(R.id.main_menu);
//        secondaryMenu = generalView.findViewById(R.id.secondary_menu);
//        secondaryMenu.setOnBackListener(v -> {
//            setModeUp(false);
//        });
//        mainMenu.setLayoutManager(new LinearLayoutManager(
//                this, LinearLayoutManager.HORIZONTAL, false
//        ));
//        ArrayList<IFLMCard> items = new ArrayList<>();
//        items.add(new PictureCardItem());
//        items.add(new SoundCardItem());
//        items.add(new InputsCardItem());
//        items.add(new RatioCardItem());
//        items.add(new TimerCardItem());
//        items.add(new NumericCardItem());
//        items.add(new SettingCardItem());
//        MainItemAdapter adapter = new MainItemAdapter(items);
//        mainMenu.setAdapter(adapter);
//
//    }
//
//    public class MainItemViewHolder extends RecyclerView.ViewHolder {
//        public final CardView card;
//        public final View actionsView;
//        public final ImageView imageView;
//        public final TextView title;
//        public final TextView subTitle;
//        public final IndicatorView progressIndicator;
//
//        MainItemViewHolder(View view) {
//            super(view);
//            card = view.findViewById(R.id.card);
//            actionsView = view.findViewById(R.id.actions_view);
//            imageView = view.findViewById(R.id.imageView);
//            title = view.findViewById(R.id.title);
//            subTitle = view.findViewById(R.id.subTitle);
//            progressIndicator = view.findViewById(R.id.progressIndicator);
//
//
////            title.getPaint().setDither(false);
////            title.getPaint().setLinearText(true);
////            title.getPaint().setStrokeWidth(0.2f);
////            title.getPaint().setStyle(Paint.Style.STROKE);
//        }
//    }
//
//    class MainItemAdapter extends RecyclerView.Adapter<MainItemViewHolder> {
//        List<IFLMCard> items;
//
//        MainItemAdapter(List<IFLMCard> items) {
//            this.items = items;
//        }
//
//        @NonNull
//        @Override
//        public MainItemViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
//            View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.menu_item, parent, false);
//            MainItemViewHolder holder = new MainItemViewHolder(view);
//
//            holder.card.setOnKeyListener((v, keyCode, event) -> {
//                if (event.getAction() == KeyEvent.ACTION_UP
//                        && event.getKeyCode() == KeyEvent.KEYCODE_DPAD_UP) {
//                    setModeUp(true);
//                    return true;
//                }
//                if (event.getAction() == KeyEvent.ACTION_UP
//                        && event.getKeyCode() == KeyEvent.KEYCODE_BACK) {
//                    stopSelf();
//                    return true;
//                }
//                return
//                        event.getKeyCode() == KeyEvent.KEYCODE_DPAD_UP ||
//                                event.getKeyCode() == KeyEvent.KEYCODE_BACK;
//            });
//            holder.card.setOnFocusChangeListener((v, hasFocus) -> {
//                holder.card.setScaleY(hasFocus ? 1.1f : 1);
//                holder.card.setScaleX(hasFocus ? 1.1f : 1);
//                holder.actionsView.setVisibility(hasFocus ? View.VISIBLE : View.INVISIBLE);
//                holder.itemView.setTranslationZ(hasFocus ? 5 : 0);
//                holder.card.setCardBackgroundColor(hasFocus ?
//                        0xFF444444 : 0xFF333333
//                );
//
//            });
//            holder.card.requestFocus();
//            return holder;
//
//        }
//
//        int i = 0;
//
//        @Override
//        public void onBindViewHolder(@NonNull MainItemViewHolder holder, int position) {
//            items.get(position).setViewHolder(holder);
//        }
//
//        @Override
//        public int getItemCount() {
//            return items.size();
//        }
//    }
//
//    private void setModeUp(boolean isTop) {
//        if (isTop) {
//            secondaryMenu.requestFocus(View.FOCUS_DOWN, null);
//        } else {
//            mainMenu.getChildAt(0).requestFocus();
//        }
//        secondaryMenu.animate().alpha(isTop ? 1 : 0).translationY(isTop ? 0 : 300).setDuration(400).start();
//        mainMenu.animate()/*.alpha(isTop ? 0.9f : 1)*/.translationY(!isTop ? 0 : 200).setDuration(400).start();
//    }
//
//
//    @Override
//    public void onDestroy() {
//        try {
//            timer.removeCallbacks(updateSleepTime);
//        } catch (Exception e) {
//
//        }
//        if (windowManager != null)
//            windowManager.removeView(generalView);
//        super.onDestroy();
//    }
//}

File: app/src/main/java/com/wezom/kiviremoteserver/service/aspect/AspectLayoutService_NEW.kt
package com.wezom.kiviremoteserver.service.aspect

import android.app.Service
import android.content.Context
import android.content.Intent
import android.graphics.Color
import android.graphics.PixelFormat
import android.os.Build
import android.os.Handler
import android.os.IBinder
import android.provider.Settings
import android.support.v4.content.ContextCompat
import android.support.v7.widget.CardView
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.KeyEvent
import android.view.View
import android.view.ViewGroup
import android.view.WindowManager
import android.view.animation.Animation
import android.view.animation.AnimationUtils
import android.view.animation.DecelerateInterpolator
import android.widget.*
import com.wezom.kiviremoteserver.App
import com.wezom.kiviremoteserver.R
import com.wezom.kiviremoteserver.environment.EnvironmentInputsHelper
import com.wezom.kiviremoteserver.environment.EnvironmentPictureSettings
import com.wezom.kiviremoteserver.service.aspect.aspect_v2.CardsMainAdapter
import com.wezom.kiviremoteserver.service.aspect.aspect_v2.InputsAdapter
import com.wezom.kiviremoteserver.service.aspect.aspect_v2.data.CardData
import com.wezom.kiviremoteserver.service.aspect.aspect_v2.data.Cards
import wezom.kiviremoteserver.environment.bridge.BridgePicture

class AspectLayoutService_NEW : Service() {

    private lateinit var pbValue: TextView
    private lateinit var pb: ProgressBar
    private lateinit var textPicker: NumberPicker
    private lateinit var windowManager: WindowManager

    private val pictureSettings: EnvironmentPictureSettings = EnvironmentPictureSettings()
    private val inputsHelper: EnvironmentInputsHelper = EnvironmentInputsHelper()

    private var autoCloseTime = 10
    private val timer = Handler()

    private val generalType = BridgePicture.LAYER_TYPE//WindowManager.LayoutParams.TYPE_TOAST;
    private lateinit var generalView: RelativeLayout
    private lateinit var mainMenu: RecyclerView
    private lateinit var inputsList: RecyclerView
    private lateinit var inputsListContainer: CardView
    private lateinit var secondaryMenu: LinearLayout
    private lateinit var backView: View

    private val updateSleepTime = object : Runnable {
        override fun run() {
            timer.postDelayed(this, 1000)
            if (autoCloseTime > 0 && System.currentTimeMillis() - autoCloseTime > lastUpdate)  {
                stopSelf()
            }
        }
    }

    override fun onCreate() {
        super.onCreate()
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M && App.checkWizard(this)) {
            return
        }

        autoCloseTime = Settings.Global.getInt(contentResolver, OSD_TIME, 10)
        if (autoCloseTime in 1..9) { autoCloseTime = 10 }

        //TODO remove autoCloseTime *= 10;
        autoCloseTime *= 1000000
        autoCloseTime *= 1000
        mainColor = resources.getColor(R.color.colorPrimary)

        lastUpdate = System.currentTimeMillis()
        initLayout(baseContext)

        timer.postDelayed(updateSleepTime, autoCloseTime.toLong())
    }

    override fun onBind(intent: Intent): IBinder? = null

    private fun initLayout(context: Context) {
        windowManager = context.applicationContext.getSystemService(Context.WINDOW_SERVICE) as WindowManager
        generalView = View.inflate(context, R.layout.layout_aspect_v2, null) as RelativeLayout
        val param = WindowManager.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.MATCH_PARENT,
                generalType, //TYPE_SYSTEM_ALERT
                WindowManager.LayoutParams.FLAG_FULLSCREEN,
                PixelFormat.TRANSLUCENT)
        windowManager.addView(generalView, param)
        generalView.visibility = View.VISIBLE
        generalView.clearAnimation()

        secondaryMenu = generalView.findViewById(R.id.secondary_menu)
        textPicker = generalView.findViewById(R.id.text_picker)
        pb = generalView.findViewById(R.id.pb)
        pbValue = generalView.findViewById(R.id.tv_value)
        backView = generalView.findViewById(R.id.back_view)

        initMainCards()
        initInputs()

        mainMenu.animate()
                .translationY(0f)
                .setDuration(400)
                .setInterpolator(DecelerateInterpolator())
                .start()

        val animFadeIn = AnimationUtils.loadAnimation(applicationContext, android.R.anim.fade_in)
        animFadeIn.duration = 600
        animFadeIn.interpolator = DecelerateInterpolator()
        backView.startAnimation(animFadeIn)
    }

    private fun initMainCards() {
        mainMenu = generalView.findViewById(R.id.main_menu)
        mainMenu.layoutManager = LinearLayoutManager(this, LinearLayoutManager.HORIZONTAL, false)
        val adapter = CardsMainAdapter(Cards.allData, { data -> setModeUp(true, data) }, {
            if (isStartedClosing) { return@CardsMainAdapter }

            isStartedClosing = true
            mainMenu.animate()/*.alpha(isTop ? 0.9f : 1)*/
                    .translationY(500f)
                    .setDuration(400)
                    .setInterpolator(DecelerateInterpolator())
                    .start()

            val animFadeOut = AnimationUtils.loadAnimation(applicationContext, android.R.anim.fade_out)
            animFadeOut.duration = 600
            animFadeOut.interpolator = DecelerateInterpolator()
            animFadeOut.setAnimationListener(object : Animation.AnimationListener {
                override fun onAnimationStart(animation: Animation) {}
                override fun onAnimationRepeat(animation: Animation) {}
                override fun onAnimationEnd(animation: Animation) { stopSelf() }
            })

            backView.startAnimation(animFadeOut)
        }, pictureSettings)
        mainMenu.adapter = adapter
    }

    private fun initInputs() {
        inputsListContainer = generalView.findViewById(R.id.cv_inputs)
        inputsList = generalView.findViewById(R.id.rv_inputs)
        inputsList.layoutManager = LinearLayoutManager(this, LinearLayoutManager.HORIZONTAL, false)
        val adapter = InputsAdapter(inputsHelper, {
            //TODO show backView
            setModeUp(false)

            val animFadeIn = AnimationUtils.loadAnimation(applicationContext, android.R.anim.fade_in)
            animFadeIn.duration = 600
            animFadeIn.interpolator = DecelerateInterpolator()
            animFadeIn.setAnimationListener(object : Animation.AnimationListener {
                override fun onAnimationStart(animation: Animation) {}
                override fun onAnimationRepeat(animation: Animation) {}
                override fun onAnimationEnd(animation: Animation) { backView.visibility = View.VISIBLE }
            })

            backView.startAnimation(animFadeIn)
        }, inputsListContainer.context)
        inputsList.adapter = adapter
    }

    var lastFocusChildIndex = 0

    private fun setModeUp(isTop: Boolean, data: CardData? = null) {
        if (isTop) {
            lastFocusChildIndex = mainMenu.indexOfChild(mainMenu.focusedChild)

            when (data?.type) {
                2 -> {
                    presentPictureModeSettings()
                    secondaryMenu.findViewById<NumberPicker>(R.id.text_picker).requestFocus()

                    secondaryMenu.animate()
                        .alpha(1.toFloat())
                        .translationY((-50).toFloat())
                        .setInterpolator(DecelerateInterpolator())
                        .setDuration(400)
                        .start()
                }
                5 -> {
                    presentSoundModeSettings()
                    secondaryMenu.findViewById<NumberPicker>(R.id.text_picker).requestFocus()

                    secondaryMenu.animate()
                            .alpha(1.toFloat())
                            .translationY((-50).toFloat())
                            .setInterpolator(DecelerateInterpolator())
                            .setDuration(400)
                            .start()
                }
                0 -> {
                    inputsListContainer.getChildAt(0).requestFocus()
                    inputsListContainer.animate()
                            .alpha(1.toFloat())
                            .translationY((-50).toFloat())
                            .setInterpolator(DecelerateInterpolator())
                            .setDuration(400)
                            .start()

                    val animFadeOut = AnimationUtils.loadAnimation(applicationContext, android.R.anim.fade_out)
                    animFadeOut.duration = 600
                    animFadeOut.interpolator = DecelerateInterpolator()
                    animFadeOut.setAnimationListener(object : Animation.AnimationListener {
                        override fun onAnimationStart(animation: Animation) {}
                        override fun onAnimationRepeat(animation: Animation) {}
                        override fun onAnimationEnd(animation: Animation) { backView.visibility = View.INVISIBLE }
                    })

                    backView.startAnimation(animFadeOut)
                }
            }

//            secondaryMenu.findViewById<NumberPicker>(R.id.text_picker).requestFocus()
        } else {
            mainMenu.getChildAt(lastFocusChildIndex).requestFocus()

            inputsListContainer.animate()
                    .alpha(0.toFloat())
                    .translationY(300.toFloat())
                    .setInterpolator(DecelerateInterpolator())
                    .setDuration(400)
                    .start()

            secondaryMenu.animate()
                .alpha(0.toFloat())
                .translationY(300.toFloat())
                .setInterpolator(DecelerateInterpolator())
                .setDuration(400)
                .start()
        }

//        inputsListContainer.animate()
//                .alpha((if (isTop) 1 else 0).toFloat())
//                .translationY((if (isTop) -50 else 300).toFloat())
//                .setInterpolator(DecelerateInterpolator())
//                .setDuration(400)
//                .start()

//        secondaryMenu.animate()
//                .alpha((if (isTop) 1 else 0).toFloat())
//                .translationY((if (isTop) -50 else 300).toFloat())
//                .setInterpolator(DecelerateInterpolator())
//                .setDuration(400)
//                .start()

        mainMenu.animate()
                .translationY((if (!isTop) 0 else 220).toFloat())
                .setInterpolator(DecelerateInterpolator())
                .setDuration(400)
                .start()
    }


    private fun presentPictureModeSettings() {
        val values = arrayOf(
                Pair(R.string.brightness, pictureSettings.brightness),
                Pair(R.string.contrast, pictureSettings.contrast),
                Pair(R.string.saturation, pictureSettings.saturation),
                Pair(R.string.temperature_des, pictureSettings.temperature),
                Pair(R.string.backlight, pictureSettings.backlight)
        )

        var indexOfSelectedPickerItem = 0
        pb.progress = values[indexOfSelectedPickerItem].second
        pbValue.text = values[indexOfSelectedPickerItem].second.toString()

        textPicker.maxValue = 0
        textPicker.displayedValues = values.map { getString(it.first) }.toTypedArray()
        textPicker.maxValue = values.size - 1
        textPicker.minValue = 0

        textPicker.setOnValueChangedListener { _, _, newValue ->
            if (newValue == 3) {
                pb.max = 5
                pb.progressDrawable = ContextCompat.getDrawable(this, R.drawable.gradient_progress_temperature)
            } else {
                pb.max = 100
                pb.progressDrawable = ContextCompat.getDrawable(this, R.drawable.gradient_progress)
            }

            pb.progress = values[newValue].second
            pbValue.text = values[newValue].second.toString()
            indexOfSelectedPickerItem = newValue
        }

        textPicker.setOnKeyListener { _, keyCode, event ->
            if (event.action == KeyEvent.ACTION_DOWN && keyCode == KeyEvent.KEYCODE_DPAD_RIGHT) {
                pb.progress = pb.progress + 1
                pbValue.text = pb.progress.toString()
                values[indexOfSelectedPickerItem] = values[indexOfSelectedPickerItem].copy(second = pb.progress)

                when(indexOfSelectedPickerItem) {
                    0 -> pictureSettings.brightness = pb.progress
                    1 -> pictureSettings.contrast = pb.progress
                    2 -> pictureSettings.saturation = pb.progress
                    3 -> pictureSettings.temperature = pb.progress
                    4 -> pictureSettings.setBacklight(pb.progress, this)
                }

                return@setOnKeyListener true
            }
            if (event.action == KeyEvent.ACTION_DOWN && keyCode == KeyEvent.KEYCODE_DPAD_LEFT) {
                if (indexOfSelectedPickerItem != 3 || pb.progress != 1) {
                    pb.progress = pb.progress - 1
                }
                pbValue.text = pb.progress.toString()
                values[indexOfSelectedPickerItem] = values[indexOfSelectedPickerItem].copy(second = pb.progress)

                when(indexOfSelectedPickerItem) {
                    0 -> pictureSettings.brightness = pb.progress
                    1 -> pictureSettings.contrast = pb.progress
                    2 -> pictureSettings.saturation = pb.progress
                    3 -> pictureSettings.temperature = pb.progress
                    4 -> pictureSettings.setBacklight(pb.progress, this)
                }

                return@setOnKeyListener true
            }
            if (event.action == KeyEvent.ACTION_UP && keyCode == KeyEvent.KEYCODE_BACK) {
                setModeUp(false)
                return@setOnKeyListener true
            }

            false
        }
    }

    private fun presentSoundModeSettings() {
        val values = arrayOf(
                Pair(R.string.sound_height, pictureSettings.getTrebleLevel(this)),
                Pair(R.string.sound_low, pictureSettings.getBassLevel(this)),
                Pair(R.string.sound_balance, pictureSettings.balanceLevel),
                Pair(R.string.sound_height, pictureSettings.getTrebleLevel(this)),
                Pair(R.string.sound_low, pictureSettings.getBassLevel(this)),
                Pair(R.string.sound_balance, pictureSettings.balanceLevel)
        )

        var indexOfSelectedPickerItem = 0
        pb.progress = values[indexOfSelectedPickerItem].second
        pbValue.text = values[indexOfSelectedPickerItem].second.toString()

        textPicker.maxValue = 0
        textPicker.displayedValues = values.map { getString(it.first) }.toTypedArray()
        textPicker.maxValue = values.size - 1
        textPicker.minValue = 0

        textPicker.setOnValueChangedListener { _, _, newValue ->
            pb.max = 100
            pb.progressDrawable = ContextCompat.getDrawable(this, R.drawable.gradient_progress)
            pb.progress = values[newValue].second
            pbValue.text = values[newValue].second.toString()
            indexOfSelectedPickerItem = newValue
        }

        textPicker.setOnKeyListener { _, keyCode, event ->
            if (event.action == KeyEvent.ACTION_DOWN && keyCode == KeyEvent.KEYCODE_DPAD_RIGHT) {
                pb.progress = pb.progress + 1
                pbValue.text = pb.progress.toString()
                values[indexOfSelectedPickerItem] = values[indexOfSelectedPickerItem].copy(second = pb.progress)

                when(indexOfSelectedPickerItem) {
                    0, 3 -> pictureSettings.setTrebleLevel(this, pb.progress)
                    1, 4 -> pictureSettings.setBassLevel(this, pb.progress)
                    2, 5 -> pictureSettings.balanceLevel = pb.progress
                }

                return@setOnKeyListener true
            }
            if (event.action == KeyEvent.ACTION_DOWN && keyCode == KeyEvent.KEYCODE_DPAD_LEFT) {
                pb.progress = pb.progress - 1
                pbValue.text = pb.progress.toString()
                values[indexOfSelectedPickerItem] = values[indexOfSelectedPickerItem].copy(second = pb.progress)

                when(indexOfSelectedPickerItem) {
                    0, 3 -> pictureSettings.setTrebleLevel(this, pb.progress)
                    1, 4 -> pictureSettings.setBassLevel(this, pb.progress)
                    2, 5 -> pictureSettings.balanceLevel = pb.progress
                }

                return@setOnKeyListener true
            }
            if (event.action == KeyEvent.ACTION_UP && keyCode == KeyEvent.KEYCODE_BACK) {
                setModeUp(false)
                return@setOnKeyListener true
            }

            false
        }
    }


    var isStartedClosing = false

//    override fun onKey(p0: View?, p1: Int, event: KeyEvent): Boolean {
//        return when {
//            event.action == KeyEvent.ACTION_UP && event.keyCode == KeyEvent.KEYCODE_DPAD_UP -> {
//                setModeUp(true)
//                true
//            }
//
//            event.action == KeyEvent.ACTION_UP && event.keyCode == KeyEvent.KEYCODE_BACK -> {
//                if (isStartedClosing) {
//                    return true
//                }
//
//                isStartedClosing = true
//                mainMenu.animate()/*.alpha(isTop ? 0.9f : 1)*/
//                        .translationY(500f)
//                        .setDuration(400)
//                        .setInterpolator(DecelerateInterpolator())
//                        .start()
//
//                val animFadeOut = AnimationUtils.loadAnimation(applicationContext, android.R.anim.fade_out)
//                animFadeOut.duration = 600
//                animFadeOut.interpolator = DecelerateInterpolator()
//                animFadeOut.setAnimationListener(object : Animation.AnimationListener {
//                    override fun onAnimationStart(animation: Animation) {}
//
//                    override fun onAnimationRepeat(animation: Animation) {}
//
//                    override fun onAnimationEnd(animation: Animation) {
//                        stopSelf()
//                    }
//                })
//
//                backView.startAnimation(animFadeOut)
//                true
//            }
//
//            else -> event.keyCode == KeyEvent.KEYCODE_DPAD_UP || event.keyCode == KeyEvent.KEYCODE_BACK
//        }
//    }

    override fun onDestroy() {
        timer.removeCallbacks(updateSleepTime)
        windowManager.removeView(generalView)
        super.onDestroy()
    }

    companion object {
        @Volatile
        var lastUpdate: Long = 0
        private var mainColor = Color.BLUE
        private val OSD_TIME = "OSD_TIME"
    }
}
